canonical/maas-ui
# refactor: machine request cache keys MAASENG-1905

## Done

- refactor: machine request cache keys MAASENG-1905
- drive-by fixes:
  - add explicit immer import (already imported as a peer dependency of redux toolkit)
  - fix machine list pagination not to rely on a re-render for the buttons to work
  - reduce duplication with enableCallIdMocks util

### Rationale
This replaces random strings used as callId with `JSON.stringify` of used request parameters. This ensures there's only a single state for each possible permutation of machin list at a time.

It also introduces `fetchedAt`, `refetchedAt` to enable time-based cache invalidation in the future.

#### before
- machine.count & machine.list are requested multiple times in quick succession on initial page load
- on machine list or count request
  - loading state is displayed when requested for the first time and items added to cache
  - loading state is displayed for subsequent requests and previously cached state is discarded
#### after
- machine.count & machine.list are requested only once on initial page load
- on machine list or count request
  - loading state is displayed when requested for the first time and items added to cache
  - no loading state is displayed for subsequent requests and previously cached state is displayed whilst an update is loading

## QA

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### QA steps

- Go to machine listing page
- Open one of the machines on the list in a new tab and go back to the previous tab
- Click to go to the next page
- Now in the second tab, perform an action on the machine, e.g. change its name, and go back to the previous tab
- Click to go to the previous page
- Ensure that no loading screen was displayed and the machine's name has been eventually updated

## Fixes

Fixes: https://warthogs.atlassian.net/browse/MAASENG-1905

## Screenshots
